### DISCUSSION TITLE
Editor Team Question: What should we do next?

### DISCUSSION THREAD

**Discussion by minecraftchloe** [Mojang/minecraft-editor#8] [Q&A]
What is the #1 _thing_ you want us to do next? Either new or iterate.

> **_As of March 2023:_** Our short-term priorities, in order, are to maintain stability in Preview, and to release v0.4: The API Update. However, you can expect to see small, incremental iterations (eg. 0.3.X) based on your feedback and our known issues. 
> 
> _v0.4: The API Update will be focused on exposing the Editor APIs for you to begin building your own Editor extensions. For now we invite you to acquaint yourself with the existing [scripting](https://docs.microsoft.com/en-us/minecraft/creator/scriptapi/) efforts and get familiar with Typescript!_

**Comment by JWForever5504** [comment]
As some people have mentioned, customizable brushes, similar to the clone command, would be amazing. 

Also, while I understand replicating the WorldEdit mod may infringe on copyright, getting as close as possible to it would be awesome, as I'm sure many of us are already familiar with it.

Random fill percentages (i.e.: randomly sprinkle some cobblestone in with stone) would be very cool, as well as being able to omit specific block types from selection, instead replacing them with a block of our choice.

This is so cool, and I can't wait to see it continue to develop!

**Reply by CaptainCooky** [reply]
This feature is an absolute must! If this editor aims to be fully functional, it eventually needs to do what WorldEdit does and do it better.

**Reply by minecraftchloe** [reply]
100% agree!! The Editor API will unlock this depth of functionality. Definitely "brush up" on your JavaScript/Typescript in preparation! :)

That said, we've got mockups for exposing multiple block selection with ratio specification + random fill as part of the next iteration of a block palette/block picker. 

Bonus fact: During an internal hackathon to test early iterations of the Editor API a while ago, a clone stamp extension (Photoshop inspired) was prototyped to stress test some other functionality.

**Comment by MisledWater79** [comment]
I made a couple discussions but like someone said in this discussion, something close to worldedit along with an API for gametest would be nice. Will add more ideas shortly 😁.

**Comment by Hatchibombotar** [comment]
How about supporting the pickblock hotkey?

**Reply by CaptainCooky** [reply]
If you are in the "Crosshair Mode", the pick block feature works as it does in Creative.

**Reply by Hatchibombotar** [reply]
Yes, but it doesn't select the block in the brush.
With a pickblock hotkey you could hover over a block with your cursor and it would be useable in brushes.

**Reply by minecraftchloe** [reply]
This is included in our mockups for the next iteration of block picker/hotbar! Stay tuned!

**Comment by guruplant** [comment]
having just started to play with it the 3 things that spring to mind are:

1, Saving different copies to use as brushes
2, being able to rotate a copy before pasting
3, having a setting that changes the paste location from the lowest x,z coordinate to the centre of the x,z coordinates

**Reply by CaptainCooky** [reply]
You can rotate the selection before committing to the paste. If you use the paste preview, (ctrl+shift+v) you can move the selection box, flip it, and rotate it. Then confirm the paste afterwards. The problem is though, there's no preview of the copied blocks, so it's like a memory game😂  
Hopefully, this feature will be improved upon.

**Comment by masonfaulk2008** [comment]
Schematics seem like a must, I design something in editor mode, and rather than porting over the whole build, I’d get a semi transparent no collision schematic, with a materials list. This would dramatically improve the survival experience and if it’s “not minecrafty enough” just simply don’t use it. A “.mcschematic“ file extension would be allowed to be imported into vanilla Minecraft for recreating in survival. The schematic would include a Materials list UI, and it’d only provide you with a holographic 3D model of your build that you can toggle on and off in the materials list UI, you have to mine and craft all of the materials, this only provides a mental advantage for the average player. I love designing my builds in Creative, but porting them over stresses me out beyond limits. A premade materials list, and a 3D schematic to move around at will, will dramatically change how a lot of people play and approach building in Minecraft. Instead of beating a dragon, these schemes can add a new personalized goal, of collecting and placing blocks to create your dream build. This is a tool, and I consider it anything but cheating, it’d be optional, but once the word gets out about it. It’d dramatically change how I and many people play, and approach Minecraft. Imagine sharing schematics with your friends or online, the joy of building that fancy new red stone contraption, or a Little cottage, with the mental advantage of a schematic. This would introduce a new way of sharing your builds and ideas to the community. And it’s quite frankly a beautiful idea, schematics will change Minecraft Bedrock for the better, they’ll be that one feature that will improve your game play experience and it’ll create a new aspect of social interactions and sharing ideas.

**Reply by CaptainCooky** [reply]
Definitely not cheating. Litematica is one such Mod on the Java version, that does exactly this.

**Comment by Landy005** [comment]
Inventory and fill percentages!
![mockup_editor](https://user-images.githubusercontent.com/16018643/223946932-6862c0e5-0c64-4236-8a0c-15debf402cf7.png)
( I forgot to put a history bar at the top, just for convenience to have your last blocks used available)

**Reply by minecraftchloe** [reply]
+1 - included already in the next iteration of block picker/palette!

**Reply by Landy005** [reply]
> +1 - included already in the next iteration of block picker/palette!

Any plans on implementing the percentage when picking a list? I think adding an inventory combined with the current searchbar would still work better than being forced to always type your block. What I also appreciate so much about fawe search suggestion is that you can arrow key up and down to go through the suggestion list, I know for instance what is at the bottom of the list when i do //set green_ and instead of going over the whole list i'd arrow up to have the last block.

![image](https://github.com/user-attachments/assets/2c87314f-44fa-4c80-9432-c5c28aa6322c)

**Reply by dacowan** [reply]
Weights are applicable once you pick a block and put it into a block list.  You can assign a "weighted block" to the block palette entry just like a normal block, but when you paint with it, it will use the weights to determine which one of the variations are placed.
This is available now

**Comment by CaptainCooky** [comment]
Looking forward to the Editor API! 
As for suggestions, it'd be nice to have an insight into the Dev team's plans for already confirmed features so we don't suggest redundant ideas. 

That said, here are my initial thoughts: 
1. More brush shapes - Sphere, Cylinder
2. Mask mode - Essential feature, only affecting specific blocks will greatly improve both brush functionality and copy/paste/fill tools. 
3. Documentation - Docs, Docs, Docs! I love reading through the documentation for these kinds of tools. It'd be nice if the officially supported docs over a Microsoft Learn were to keep up with Development. 😁

Love the Editor, good job, and I'm looking forward to the future of this tool!

**Reply by SineVector241** [reply]
Would love to see more brush shapes and maybe more brush and selection options such as a filter to exclude or replace blocks within the selection or brush. Maybe also being able to set block data that is currently available in the fill command. There is so much potential for this tool!

**Reply by minecraftchloe** [reply]
Some answers now that we sort of have some to share! :D 

1. Additional brush shapes: YES! We want to add more shapes to the core tool, but also it will be possible to create your own shapes with the Editor API. Core tool shapes we're thinking of adding (with parameters for height/depth/etc. depending on shape): Cube (already there), Cylinder, Sphere, Pyramid, and Rectangle _(this would be custom specification of length/width/height, so theoretically if all were equal it'd be_ like _a cube_). 
2. Mask mode: Absolutely on the docket - we've got mocks of how this would be implemented. We just have to implement it! 
3. Documentation: Will be posted on Microsoft Learn! API documentation will exist in the Reference Documentation area, and we're going to add a high-level general article around Editor Extensions in the Creator Tools -> Bedrock Editor area.

**Comment by cute-catgirl** [comment]
A "play" mode to test your map, that will basically clone the world (all changes made in play mode won't be saved) and allow you to test it as a user, similar to how it works in games like Mario Maker

Edit: also a way to export as a java world, having something like Chunker built in

**Reply by minecraftchloe** [reply]
100% yes for "playtesting" your map in a way that isn't super time consuming and bulky! Internally we're referring to this as "Edit/Test," and we've moved the priority of this up, and we're hoping to expose a "beta" version of this. Stay tuned! 

Having something like Chunker built-in: Major wishlist item on our side, too! Some dependencies for us to work through, but know it's on our radar! <3

**Comment by TheMightyDarkness** [comment]
There could be a way to edit biome data. An absolute need, because Bedrock does not have /fillbiome, unlike Java.

**Reply by bryhaw** [reply]
It's on the backlog!

**Comment by StealthyExpertX** [comment]
I’d love to see more low-level access to world and pack data to enable plugins using the API. Features like rendering custom entities or items, blocks from installed packs and accessing raw data from .mcstructure and .nbt files would be great!

Accessing block model data like meshes and texture atlas would be cool too!
